File: Тренировки по алгоритмам 3.0/Тема 0. Разминка/03.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <utility>

using namespace std;

int stickers_count(const vector<int>& diego, int k) {
	auto it = lower_bound(diego.begin(), diego.end(), k);
	if (it == diego.end()) {
		return diego.size();
	}
	if (it == diego.begin()) {
		return *it < k ? 1 : 0;
	}
	return distance(diego.begin(), it);
}

int main()
{
	int n = 0;
	cin >> n;
	vector<int> diego_raw;
	diego_raw.resize(n);
	for (int i = 0; i < n; ++i) {
		cin >> diego_raw[i];
	}
	int k = 0;
	cin >> k;
	vector<int> collectors;
	collectors.resize(k);
	for (int i = 0; i < k; ++i) {
		cin >> collectors[i];
	}
	unordered_set<int> diego_set(diego_raw.begin(), diego_raw.end());
	vector<int> diego(diego_set.begin(), diego_set.end());
	sort(diego.begin(), diego.end());
	for (int k : collectors) {
		cout << stickers_count(diego, k) << '\n';
	}
}

File: Тренировки по алгоритмам 3.0/Тема 0. Разминка/03.py
def lower_bound(arr, target, begin, end):
    if end - begin < 2:
        return begin if target <= arr[begin] else end
    mid = (begin + end) // 2
    if arr[mid] == target:
        return mid
    if target < arr[mid]:
        return lower_bound(arr, target, begin, mid)
    else:
        return lower_bound(arr, target, mid, end)


def stickers_count(diego, k, size) -> int:
    i = lower_bound(diego, k, 0, size)
    if i == len(diego):
        return i
    if i == 0:
        return 1 if diego[i] < k else 0
    return i


def main():
    n = int(input())
    diego = sorted(set([int(x) for x in input().split()]))
    k = int(input())
    size = len(diego)
    kl = [int(x) for x in input().split()]
    for k in kl:
        print(stickers_count(diego, k, size))


if __name__ == '__main__':
    main()
